xwayland/selection.c

```diff
@@ -511,33 +596,83 @@ static void xwm_get_selection_targets(struct wlr_xwm *xwm) {
 
 		if (mime_type != NULL) {
 			char **mime_type_ptr =
-				wl_array_add(&source->base.mime_types, sizeof(*mime_type_ptr));
+				wl_array_add(mime_types, sizeof(*mime_type_ptr));
 			if (mime_type_ptr == NULL) {
 				break;
 			}
 			*mime_type_ptr = mime_type;
 
 			xcb_atom_t *atom_ptr =
-				wl_array_add(&source->mime_types_atoms, sizeof(*atom_ptr));
+				wl_array_add(mime_types_atoms, sizeof(*atom_ptr));
 			if (atom_ptr == NULL) {
 				break;
 			}
 			*atom_ptr = value[i];
 		}
 	}
 
-	wlr_seat_set_selection(xwm->seat, &source->base,
-		wl_display_next_serial(xwm->xwayland->wl_display));
-
 	free(reply);
+	return true;
+}
+
+static void xwm_selection_get_targets(struct wlr_xwm_selection *selection) {
+	// set the wayland selection to the X11 selection
+	struct wlr_xwm *xwm = selection->xwm;
+
+	if (selection == &xwm->clipboard_selection) {
+		struct x11_data_source *source =
+			calloc(1, sizeof(struct x11_data_source));
+		if (source == NULL) {
+			return;
+		}
+		wlr_data_source_init(&source->base);
+		source->base.accept = data_source_accept;
+		source->base.send = data_source_send;
+		source->base.cancel = data_source_cancel;
+
+		source->selection = selection;
+		wl_array_init(&source->mime_types_atoms);
+
+		bool ok = source_get_targets(selection, &source->base.mime_types,
+			&source->mime_types_atoms);
+		if (ok) {
+			wlr_seat_set_selection(xwm->seat, &source->base,
+				wl_display_next_serial(xwm->xwayland->wl_display));
```

Comment:
Setting the selection like this with a generated serial will cause some problems with #511. There's no way for the seat to internally verify this serial. We'll probably need to have a way to add serials to the validation list that we know to be valid internally.